ambry-cloud/src/main/java/com/github/ambry/cloud/azure/AzureStorageCompactor.java

```diff
@@ -257,17 +289,67 @@ int purgeBlobs(List<CloudBlobMetadata> blobMetadataList) throws CloudStorageExce
   }
 
   /**
-   * Returns the dead blob in the specified partition with the earliest expiration or deletion time.
+   * Get the current compaction progress for a partition.
    * @param partitionPath the partition to check.
-   * @param fieldName the field name to use (expiration or deletion time).
-   * @return the {@link CloudBlobMetadata} for the dead blob, or NULL if none was found.
-   * @throws CloudStorageException
+   * @return a {@link Map} containing the progress time for compaction based on deletion and expiration time.
+   * @throws CloudStorageException if the operation fails.
+   */
+  Map<String, Long> getCompactionProgress(String partitionPath) throws CloudStorageException {
+    ByteArrayOutputStream baos = new ByteArrayOutputStream(CHECKPOINT_BUFFER_SIZE);
+    boolean hasCheckpoint = requestAgent.doWithRetries(
+        () -> azureBlobDataAccessor.downloadFile(CHECKPOINT_CONTAINER, partitionPath, baos, false),
+        "Download compaction checkpoint", partitionPath);
+    if (!hasCheckpoint) {
+      return emptyCheckpoints;
+    }
+    try {
+      // Payload format: {"expirationTime" : 12345, "deletionTime" : 67890}
+      ObjectNode jsonNode = (ObjectNode) objectMapper.readTree(baos.toString());
+      Map<String, Long> checkpoints = new HashMap<>();
+      for (String fieldName : new String[]{CloudBlobMetadata.FIELD_DELETION_TIME,
+          CloudBlobMetadata.FIELD_EXPIRATION_TIME}) {
+        checkpoints.put(fieldName, jsonNode.has(fieldName) ? jsonNode.get(fieldName).longValue() : DEFAULT_TIME);
```

Comment:
For me, it seems cleaner for checkpoints to be a simple object (instead of a map) with getDeletionTimeCheckpoint and getFieldExpirationTime and toBytes method so the logic around serde/defaults can all be in one place.

Reply:
Good idea.  I'll change it.